### PR TITLE
KAFKA-7317: Use collections subscription for main consumer to reduce metadata

### PR DESCRIPTION
**Pull request by ableegoldman**
Also addresses [KAFKA-8821](https://issues.apache.org/jira/browse/KAFKA-8821)

Note that we still have to fall back to using pattern subscription if the user has added any regex-based source nodes to the topology. Includes some minor cleanup on the side

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 1]
```diff
@@ -87,10 +85,6 @@
     // map from sink processor names to sink topic (without application-id prefix for internal topics)
     private final Map<String, String> nodeToSinkTopic = new HashMap<>();
 
-    // map from topics to their matched regex patterns, this is to ensure one topic is passed through on source node
-    // even if it can be matched by multiple regex patterns
-    private final Map<String, Pattern> topicToPatterns = new HashMap<>();
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:92]: This is just moved since it's only used by a single class

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 2]
```diff
@@ -1048,24 +1047,23 @@ private void buildProcessorNode(final Map<String, ProcessorNode> processorMap,
...
-                    stringPatternEntry.getKey(),
-                    sourceNode.getTopics(subscriptionUpdates.getUpdates()));
-                log.debug("nodeToSourceTopics {}", nodeToSourceTopics);
+        if (hasSubscriptionUpdates()) {
+            for (final String nodeName : nodeToSourcePatterns.keySet()) {
+                //need to update nodeToSourceTopics and sourceTopicNames with topics matched from given regex
+                final List<String> sourceTopics = ((SourceNodeFactory) nodeFactories.get(nodeName))
+                                                      .getTopics(subscriptionUpdates);
+                nodeToSourceTopics.put(nodeName, sourceTopics);
+                sourceTopicNames.addAll(sourceTopics);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1067]: Also update `sourceTopicNames` to keep in sync with `nodeToSourceTopics`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 3]
```diff
@@ -1200,37 +1198,28 @@ private String decorateTopic(final String topic) {
...
+        log.debug("No source topics using pattern subscription found, using regular subscription for the main consumer.");
+
+        return sourceTopicNames;
     }
 
     synchronized Pattern sourceTopicPattern() {
+        log.debug("Found pattern subscribed source topics, falling back to pattern subscription for the main consumer.");
+
         if (topicPattern == null) {
-            final List<String> allSourceTopics = new ArrayList<>();
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1218]: This can be simplified by just using `sourceTopicNames`, which is identical to `nodeToSourceTopics.values()` but without the global topics, which we remove

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 4]
```diff
@@ -1856,42 +1845,25 @@ private static String nodeNames(final Set<TopologyDescription.Node> nodes) {
...
-        final SubscriptionUpdates subscriptionUpdates = new SubscriptionUpdates();
-        log.debug("{}found {} topics possibly matching regex", logPrefix, topics);
-        // update the topic groups with the returned subscription set for regex pattern subscriptions
-        subscriptionUpdates.updateTopics(topics);
-        updateSubscriptions(subscriptionUpdates, logPrefix);
+    boolean hasSubscriptionUpdates() {
+        return !subscriptionUpdates.isEmpty();
     }
 
+    synchronized void updateSubscribedTopics(final Set<String> topics,
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1872]: The `SubscriptionUpdates` class and various `updateXXX` methods were unnecessarily complex given they all boiled down to just the 4 lines of actual actions below

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 5]
```diff
@@ -717,7 +717,15 @@ private void runLoop() {
 
     private void enforceRebalance() {
         consumer.unsubscribe();
-        consumer.subscribe(builder.sourceTopicPattern(), rebalanceListener);
+        subscribeConsumer();
+    }
+
+    private void subscribeConsumer() {
+        if (builder.usesPatternSubscription()) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:720]: This is the actual fix; if the user has not themselves added pattern source topics we will go back to using regular subscription (having safely disabled auto topic creation)

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 6]
```diff
@@ -1064,6 +1064,9 @@ private void checkIfUnexpectedUserSpecifiedConsumerConfig(final Map<String, Obje
         consumerProps.put(ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG, StreamsPartitionAssignor.class.getName());
         consumerProps.put(WINDOW_STORE_CHANGE_LOG_ADDITIONAL_RETENTION_MS_CONFIG, getLong(WINDOW_STORE_CHANGE_LOG_ADDITIONAL_RETENTION_MS_CONFIG));
 
+        // disable auto topic creation
+        consumerProps.put(ConsumerConfig.ALLOW_AUTO_CREATE_TOPICS_CONFIG, "false");
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:1068]: qq - what does this mean for users that have enabled auto topic creation? Although it's not a best practice, this seems it could lead to unexpected behavior.

**Review by bbejeck — COMMENTED** [review]
Thanks for the PR @ableegoldman.  I've made a quick pass.  
The build didn't run but when I ran the tests locally I get some checkstyle failures for `InternalTopologyBuilder.`

**Comment by bbejeck** [comment]
Retest this please.

**Comment by bbejeck** [comment]
Strange, the PR builder didn't run on PR submission but I was able to trigger the build with "Retest..."

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 6 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:1068]: Do you mean on the server side or on the client? I agree we should consider checking for this in the user supplied configs before overriding it, but I think the reasonable _default_ behavior is to disable it (by client config, since it's enabled server-side by default).
If users really do want it then it's on them to enable it through the main consumer config. On the other hand, prior to this it was effectively disabled  permanently by the workaround of using pattern subscription. WDYT?

Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:1068]: Yeah, that SGTM I forgot that auto topic creation is the default server-side.

**Comment by bbejeck** [comment]
@ableegoldman failures are related - checkstyle failures.

**Comment by bbejeck** [comment]
retest this please

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilderTest.java [hunk 7]
```diff
@@ -667,30 +680,25 @@ public void shouldAddInternalTopicConfigForRepartitionTopics() {
 
     @SuppressWarnings("unchecked")
     @Test
-    public void shouldSetCorrectSourceNodesWithRegexUpdatedTopics() throws Exception {
+    public void shouldSetCorrectSourceNodesWithRegexUpdatedTopics() {
         builder.addSource(null, "source-1", null, null, null, "topic-foo");
         builder.addSource(null, "source-2", null, null, null, Pattern.compile("topic-[A-C]"));
         builder.addSource(null, "source-3", null, null, null, Pattern.compile("topic-\\d"));
 
-        final InternalTopologyBuilder.SubscriptionUpdates subscriptionUpdates = new InternalTopologyBuilder.SubscriptionUpdates();
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilderTest.java:675]: Nice to get rid of the use of reflection in the tests here and below

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 8]
```diff
@@ -215,6 +212,10 @@ Processor describe() {
         }
     }
 
+    // Map from topics to their matched regex patterns, this is to ensure one topic is passed through on source node
+    // even if it can be matched by multiple regex patterns. Only used by SourceNodeFactory
+    private final Map<String, Pattern> topicToPatterns = new HashMap<>();
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:220]: I think in a previous commit this was `static` and I had concerns that it wouldn't work, so I'm glad to see you changed this.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 9]
```diff
@@ -1048,24 +1049,23 @@ private void buildProcessorNode(final Map<String, ProcessorNode> processorMap,
...
-                    (SourceNodeFactory) nodeFactories.get(stringPatternEntry.getKey());
-                //need to update nodeToSourceTopics with topics matched from given regex
-                nodeToSourceTopics.put(
-                    stringPatternEntry.getKey(),
-                    sourceNode.getTopics(subscriptionUpdates.getUpdates()));
-                log.debug("nodeToSourceTopics {}", nodeToSourceTopics);
+        if (hasSubscriptionUpdates()) {
+            for (final String nodeName : nodeToSourcePatterns.keySet()) {
+                //need to update nodeToSourceTopics and sourceTopicNames with topics matched from given regex
+                final List<String> sourceTopics = ((SourceNodeFactory) nodeFactories.get(nodeName))
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: nit: I understand the change here for compactness, but I find it a little hard to follow.  This is subjective however so feel free to keep as is.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java (hunk 2 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1067]: Do we need to do this at this point?  I guess so at it makes sense to have `sourceTopicNames` match what's in `nodeToSourceTopics`.  I'm only asking as we never had this before and I'm curious as to why.

**Review by bbejeck — APPROVED** [review]
Thanks for this fix @ableegoldman.  LGTM, I only have minor comments.  We need to get a build running before merging though.

**Comment by bbejeck** [comment]
Retest this please.

**Comment by bbejeck** [comment]
Retest this please.

**Comment by bbejeck** [comment]
Java 8 build https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/321/
Java 11 build https://builds.apache.org/job/kafka-pr-jdk11-scala2.13/4333/

~For some reason GitHub page not always updating when build is kicked off~

EDIT: Builds just showed up now. ???

**Comment by ableegoldman** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java (hunk 9 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: I kind of thought this way was easier to understand, but I did go back and forth on it. I'm happy to change it back

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1067]: Well I did switch `#sourceTopicPattern` and the new `#sourceTopicCollection` to use `sourceTopicNames` instead of `nodeToSourceTopics.values`, in which case we do still need to do this.
It shouldn't matter, although I will say the hardest part of working with this code/class was figuring out which data structures did/did not contain what contents or updates. I'm inclined to leave this in so that the two similar data structures are kept in sync, but I'm fine with removing it just to keep the changes minimal/necessary

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilderTest.java (hunk 7 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilderTest.java:675]: 😄Agreed!

**Comment by bbejeck** [comment]
Retest this please.

**Comment by bbejeck** [comment]
Java 8 build - https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/326/

**Comment by bbejeck** [comment]
Java 8 passed
Java 11 failed with `Build step 'Execute shell' marked build as failure`

Retest this please.

**Comment by bbejeck** [comment]
Retest this please.

**Comment by bbejeck** [comment]
Java 8 failed with 
```
20:39:56 # There is insufficient memory for the Java Runtime Environment to continue.
20:39:56 # Cannot create GC thread. Out of system resources.
```

Java 11 passed https://builds.apache.org/view/All/job/kafka-pr-jdk11-scala2.13/4355/

**Comment by bbejeck** [comment]
Java 11 failed with `kafka.admin.DeleteConsumerGroupsTest.testDeleteCmdAllGroups`
Java 8 passed

**Comment by bbejeck** [comment]
Retest this please.

**Comment by bbejeck** [comment]
Java 8 on previous PR build passed.  On this run  it failed on an environmental issue ` Caused by: java.lang.OutOfMemoryError: unable to create new native thread`

Java 11 passed this run.

So I'm merging this PR.

**Comment by bbejeck** [comment]
Merged #7969 into trunk.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 10]
```diff
@@ -1865,42 +1861,25 @@ private static String nodeNames(final Set<TopologyDescription.Node> nodes) {
...
-        // update the topic groups with the returned subscription set for regex pattern subscriptions
-        subscriptionUpdates.updateTopics(topics);
-        updateSubscriptions(subscriptionUpdates, logPrefix);
+    boolean hasSubscriptionUpdates() {
+        return !subscriptionUpdates.isEmpty();
     }
 
+    synchronized void updateSubscribedTopics(final Set<String> topics,
+                                             final String logPrefix) {
+        log.debug("{}found {} topics possibly matching subscription", logPrefix, topics);
```
Comment (tedyu) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1874]: nit: missing space between logPrefix and 'found'

Comment (tedyu) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1874]: topics is a Set.
What's your intention for the second parameter ?
If you want the number of topics logged, you should use topics.size().

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 11]
```diff
@@ -1865,42 +1861,25 @@ private static String nodeNames(final Set<TopologyDescription.Node> nodes) {
...
+        subscriptionUpdates.clear();
+        subscriptionUpdates.addAll(topics);
+
+        log.debug("{}updating builder with {} topic(s) with possible matching regex subscription(s)",
```
Comment (tedyu) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1878]: Similar comment as above two.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java (hunk 10 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1874]: The spacing kept bothering me too but it's actually correct (logPrefix has a space at the end) -- we should actually clean up this class at some point to use `log = new LogContext(logPrefix).logger(getClass());` rather than explicitly insert the prefix everywhere.
I didn't write this log message, just moved things around, but I think the intention was to list the actual topics not just the size. I probably could've improved the wording of it though

Comment (tedyu) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1874]: Since topics Set can be quite large, I doubt the intention was to show the contents.
'{} topics' reads like the count of entries should be shown.

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1874]: Well the topics set should only include the matching topics that the Streams app is actually subscribed to, but I agree the wording suggests it should be the size/count. Since this PR is already merged do you want to submit a quick follow-up PR?

Comment (tedyu) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1874]: Created https://github.com/apache/kafka/pull/8005
